src/test/java/com/yelp/nrtsearch/server/grpc/LuceneServerTest.java

```diff
@@ -604,15 +608,93 @@ public void testDeleteIndex() throws IOException, InterruptedException {
     assertEquals(2, statsResponse.getNumDocs());
     assertEquals(2, statsResponse.getMaxDoc());
 
+    String indexName = "test_index";
     // deleteIndex
     DeleteIndexRequest deleteIndexRequest =
-        DeleteIndexRequest.newBuilder().setIndexName("test_index").build();
+        DeleteIndexRequest.newBuilder().setIndexName(indexName).build();
     DeleteIndexResponse deleteIndexResponse =
         grpcServer.getBlockingStub().deleteIndex(deleteIndexRequest);
 
+    Path indexRootDir = Paths.get(grpcServer.getIndexDir(), indexName);
+    assertEquals(false, Files.exists(indexRootDir));
+
     assertEquals("ok", deleteIndexResponse.getOk());
   }
 
+  /**
+   * This test creates the index, deletes the same index and then creates the same index again. This
+   * verifies whether the global state is correctly updated, flushed to disk and the previous global
+   * states are deleted.
+   */
+  @Test
+  public void testReCreateDeletedIndex() throws IOException {
+    String indexName = "test_idx_1";
+    LuceneServerGrpc.LuceneServerBlockingStub blockingStub = grpcServer.getBlockingStub();
+
+    GlobalState globalState = grpcServer.getGlobalState();
+    // verify that globalState has no content
+    assertEquals(0, Files.list(globalState.getStateDir()).count());
+
+    CreateIndexRequest createIndexRequest =
+        CreateIndexRequest.newBuilder().setIndexName(indexName).build();
+    CreateIndexResponse createIndexResponse = blockingStub.createIndex(createIndexRequest);
+    assertEquals(
+        String.format("Created Index name: %s", indexName, grpcServer.getIndexDir()),
+        createIndexResponse.getResponse());
+
+    // verify that only indices.0 exists and has the same content on disk as in-memory
+    assertEquals(1, Files.list(globalState.getStateDir()).count());
+    assertEquals(
+        "indices.0",
+        Files.list(globalState.getStateDir()).findAny().get().getFileName().toString());
+
+    Path primaryStateIndexPath = globalState.getStateDir().resolve("indices.0");
+    JsonObject persistedIndexNames =
+        JsonParser.parseString(Files.readString(primaryStateIndexPath)).getAsJsonObject();
+
+    assertEquals(
+        globalState.getIndexNames(), Collections.unmodifiableSet(persistedIndexNames.keySet()));
```

Comment:
Do we need to call `Collections.unmodifiableSet` on the set before asserting?

Reply:
Good catch, we don't, above statements are now updated.